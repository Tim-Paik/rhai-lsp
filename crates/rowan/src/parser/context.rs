//! The parser context is a separate module to limit
//! the API surface for the parser functions.
#![allow(dead_code)]
use rowan::{GreenNodeBuilder, TextRange, TextSize};

use crate::syntax::{
    Lexer,
    SyntaxKind::{self, *},
};

use super::{Parse, ParseError, ParseErrorKind};

pub(crate) struct Context<'src> {
    lexer: Lexer<'src>,
    current_token: Option<SyntaxKind>,
    last_token: Option<SyntaxKind>,
    green: GreenNodeBuilder<'static>,
    errors: Vec<ParseError>,

    /// Tracks statements being separated by ";".
    statement_closed: bool,
}

impl<'src> Context<'src> {
    pub(crate) fn new(source: &'src str) -> Self {
        Self {
            lexer: Lexer::new(source),
            current_token: None,
            last_token: None,
            green: GreenNodeBuilder::new(),
            errors: Vec::new(),

            statement_closed: true,
        }
    }

    pub(crate) fn token(&mut self) -> Option<SyntaxKind> {
        // Eat insignificant tokens
        loop {
            if self.current_token.is_none() {
                self.current_token = self.lexer.next();
            }

            match self.current_token {
                Some(COMMENT_BLOCK) | Some(COMMENT_LINE) | Some(WHITESPACE) => {
                    self.eat();
                    self.last_token = None // We aren't interested in these.
                }
                Some(ERROR) => {
                    self.eat_error(ParseErrorKind::InvalidInput);
                    self.last_token = None // We aren't interested in these.
                }
                _ => break,
            }
        }

        self.current_token
    }

    pub(crate) fn previous_token(&self) -> Option<SyntaxKind> {
        self.last_token
    }

    pub(crate) fn eat(&mut self) {
        if let Some(t) = self.current_token.take() {
            self.green.token(t.into(), self.lexer.slice());
            self.last_token = Some(t);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_as(&mut self, kind: SyntaxKind) {
        if self.current_token.is_some() {
            self.green.token(kind.into(), self.lexer.slice());
            self.last_token = Some(kind);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_error(&mut self, error: ParseErrorKind) {
        self.add_error(error);
        self.eat();
    }

    pub(crate) fn eat_error_as(&mut self, kind: SyntaxKind, error: ParseErrorKind) {
        self.add_error(error);
        self.eat_as(kind);
    }

    pub(crate) fn add_error(&mut self, error: ParseErrorKind) {
        let span = self.lexer.span();
        self.errors.push(ParseError::new(
            TextRange::new(
                TextSize::from(span.start as u32),
                TextSize::from(span.end as u32),
            ),
            error,
        ));
    }

    pub(crate) fn start_node(&mut self, kind: SyntaxKind) {
        self.green.start_node(kind.into());
    }

    pub(crate) fn finish_node(&mut self) {
        self.green.finish_node();
    }

    pub(crate) fn finish(self) -> Parse {
        Parse {
            errors: self.errors,
            green: self.green.finish(),
        }
    }

    pub(crate) fn statement_closed(&self) -> bool {
        self.statement_closed
    }

    pub(crate) fn set_statement_closed(&mut self, v: bool) {
        self.statement_closed = v;
    }
}
