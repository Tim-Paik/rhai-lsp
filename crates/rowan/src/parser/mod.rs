use crate::syntax::{SyntaxKind, SyntaxNode};
use rowan::GreenNode;
use thiserror::Error;

use self::context::Context;

mod context;
pub mod parsers;

pub struct Parser<'src> {
    context: Context<'src>,
}

impl<'src> Parser<'src> {
    pub fn new(source: &'src str) -> Self {
        Self {
            context: Context::new(source),
        }
    }

    pub fn finish(self) -> Parse {
        self.context.finish()
    }
}

impl<'src> Parser<'src> {
    pub fn execute<F: FnOnce(&mut Context)>(&mut self, f: F) {
        f(&mut self.context)
    }
}

/// The result of parsing.
#[derive(Debug, Clone)]
pub struct Parse {
    pub errors: Vec<ParseError>,
    pub green: GreenNode,
}

impl Parse {
    pub fn into_syntax(self) -> SyntaxNode {
        SyntaxNode::new_root(self.green)
    }
}

#[derive(Debug, Clone, Error)]
#[error("parse error at {range:?}: {kind}")]
pub struct ParseError {
    pub range: rowan::TextRange,
    pub kind: ParseErrorKind,
}

impl ParseError {
    pub fn new(range: rowan::TextRange, kind: ParseErrorKind) -> Self {
        Self { range, kind }
    }
}

#[derive(Debug, Clone, Error)]
pub enum ParseErrorKind {
    #[error("unexpected EOF")]
    UnexpectedEof,

    #[error(r#"invalid input"#)]
    InvalidInput,

    #[error(r#"unexpected token"#)]
    UnexpectedToken,

    #[error(r#"expected token "{0:?}""#)]
    ExpectedToken(SyntaxKind),

    #[error(
        r#"expected one of "{}""#,
        .0
            .iter()
            .map(|s|format!("{:?}", s))
            .intersperse(String::from(", "))
            .collect::<String>()
    )]
    ExpectedOneOfTokens(Vec<SyntaxKind>),
}
