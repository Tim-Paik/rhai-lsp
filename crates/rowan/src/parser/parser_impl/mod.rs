#![deny(unreachable_patterns)]

use super::context::Context;
use crate::parser::ParseErrorKind;
use crate::syntax::SyntaxKind::*;
use crate::T;

macro_rules! require_token {
    ($ctx:ident) => {
        match $ctx.token() {
            Some(t) => t,
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

    ($ctx:ident, finish_node) => {
        match $ctx.token() {
            Some(t) => t,
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::UnexpectedEof);
                return;
            }
        }
    };
}

macro_rules! expect_token {
    ($ctx:ident, finish_node, $($token:tt)*) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => {
                    $ctx.finish_node();
                    $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                    return;
                }
            }
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                return;
            }
        }
    };
    ($ctx:ident, $($token:tt)+) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
            }
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

}

impl<'src> super::Parser<'src> {
    pub fn parse_file(&mut self) {
        parse_file(&mut self.context);
    }
}

fn parse_file(ctx: &mut Context) {
    ctx.start_node(FILE);
    if let Some(SHEBANG) = ctx.token() {
        parse_shebang(ctx);
    }

    ctx.set_statement_closed(true);
    while ctx.token().is_some() {
        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }
    ctx.set_statement_closed(true);

    ctx.finish_node()
}

fn parse_shebang(ctx: &mut Context) {
    let token = require_token!(ctx);

    if !matches!(token, SHEBANG) {
        return ctx.eat_error(ParseErrorKind::ExpectedToken(SHEBANG));
    }

    ctx.eat()
}

fn parse_stmt(ctx: &mut Context) {
    let token = require_token!(ctx);
    ctx.start_node(STMT);
    ctx.set_statement_closed(false);

    if token == T![";"] {
        ctx.finish_node();
        ctx.set_statement_closed(true);
        return ctx.eat();
    }

    parse_item(ctx);

    if let Some(token) = ctx.token() {
        if token == T![";"] {
            ctx.eat();
            ctx.set_statement_closed(true);
        }
    }

    ctx.finish_node();
}

fn parse_item(ctx: &mut Context) {
    ctx.start_node(ITEM);

    // Parse doc comments if any.
    while matches!(
        require_token!(ctx, finish_node),
        COMMENT_BLOCK_DOC | COMMENT_LINE_DOC
    ) {
        ctx.eat();
    }

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr(ctx: &mut Context) {
    ctx.start_node(EXPR);
    let token = require_token!(ctx, finish_node);

    match token {
        T!["let"] => parse_expr_let(ctx),
        T!["const"] => parse_expr_const(ctx),
        T!["{"] => parse_expr_block(ctx),
        T!["fn"] => parse_expr_fn(ctx),
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => parse_expr_lit(ctx),
        IDENT => parse_expr_ident(ctx),
        _ => ctx.eat_error(ParseErrorKind::UnexpectedToken),
    }

    ctx.finish_node();
}

fn parse_expr_ident(ctx: &mut Context) {
    ctx.start_node(EXPR_IDENT);
    parse_ident(ctx);
    ctx.finish_node();
}

fn parse_expr_lit(ctx: &mut Context) {
    ctx.start_node(EXPR_LIT);
    parse_lit(ctx);
    ctx.finish_node();
}

fn parse_expr_let(ctx: &mut Context) {
    ctx.start_node(EXPR_LET);

    expect_token!(ctx, finish_node, T!["let"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_const(ctx: &mut Context) {
    ctx.start_node(EXPR_CONST);

    expect_token!(ctx, finish_node, T!["const"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_block(ctx: &mut Context) {
    ctx.start_node(EXPR_BLOCK);

    expect_token!(ctx, finish_node, T!["{"]);

    ctx.set_statement_closed(true);
    loop {
        let token = require_token!(ctx, finish_node);

        if token == T!["}"] {
            break;
        }

        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }

    expect_token!(ctx, finish_node, T!["}"]);

    // Blocks also don't require ";" when used as statements.
    ctx.set_statement_closed(true);

    ctx.finish_node();
}

fn parse_expr_fn(ctx: &mut Context) {
    ctx.start_node(EXPR_FN);

    expect_token!(ctx, finish_node, T!["fn"]);
    expect_token!(ctx, finish_node, T!["ident"]);

    parse_param_list(ctx);
    parse_expr_block(ctx);

    ctx.finish_node();
}

fn parse_param_list(ctx: &mut Context) {
    ctx.start_node(PARAM_LIST);

    let start_token = require_token!(ctx, finish_node);

    match start_token {
        T!["("] | T!["|"] => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![T!["("], T!["|"]]));
            return;
        }
    }

    loop {
        let token = require_token!(ctx, finish_node);
        if matches!(
            (start_token, token),
            (T!["("], T![")"]) | (T!["|"], T!["|"])
        ) {
            ctx.eat();
            break;
        }

        parse_param(ctx);

        let end_token = require_token!(ctx, finish_node);

        match (start_token, end_token) {
            (T!["("], T![")"]) | (T!["|"], T!["|"]) => {
                ctx.eat();
                break;
            }
            (_, T![","]) => {
                ctx.eat();
            }
            _ => {
                ctx.eat_error(ParseErrorKind::ExpectedToken(T![","]));
                break;
            }
        }
    }

    ctx.finish_node();
}

fn parse_param(ctx: &mut Context) {
    ctx.start_node(PARAM);

    expect_token!(ctx, finish_node, T!["ident"]);

    ctx.finish_node();
}

fn parse_lit(ctx: &mut Context) {
    ctx.start_node(LIT);
    let token = require_token!(ctx, finish_node);

    match token {
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![
                LIT_INT, LIT_FLOAT, LIT_BOOL, LIT_STR, LIT_CHAR,
            ]));
        }
    }

    ctx.finish_node();
}

fn parse_ident(ctx: &mut Context) {
    let token = require_token!(ctx);

    match token {
        IDENT => ctx.eat(),
        _ => ctx.eat_error(ParseErrorKind::ExpectedToken(IDENT)),
    }
}

// Expr =
// | ExprUni
// | ExprBin
// | ExprParen
// | ExprArray
// | ExprIndex
// | ExprObject
// | ExprCall
// | ExprMethodCall
// | ExprAccess
// | ExprClosure
// | ExprIf
// | ExprLoop
// | ExprFor
// | ExprWhile
// | ExprBreak
// | ExprContinue
// | ExprSwitch
// | ExprReturn
// | ExprPath
// | ExprImport
