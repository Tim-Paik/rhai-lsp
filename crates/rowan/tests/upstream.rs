use rhai_rowan::parser::Parser;
use test_case::test_case;

#[test_case(include_str!("../../../rhai/scripts/array.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/assignment.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/comments.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/fibonacci.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/for2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/function_decl4.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/if2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/loop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/mat_mul.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/module.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/oop.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op1.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op2.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/op3.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/primes.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/speed_test.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/string.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/strings_map.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/switch.rhai"))]
#[test_case(include_str!("../../../rhai/scripts/while.rhai"))]
fn parse_rhai_upstream(src: &str) {
    let parse = Parser::new(src).parse_script();
    assert!(parse.errors.is_empty());
    // TODO(tests)
    // insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));
}
