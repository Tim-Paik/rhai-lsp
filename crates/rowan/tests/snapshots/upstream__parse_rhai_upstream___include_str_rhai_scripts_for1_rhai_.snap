---
source: crates/rowan/tests/upstream.rs
expression: "format!(\"{:#?}\", parse.into_syntax())"

---
FILE@0..486
  COMMENT_LINE@0..29 "// This script runs f ..."
  WHITESPACE@29..31 "\n\n"
  STMT@31..76
    ITEM@31..75
      EXPR@31..75
        EXPR_LET@31..75
          KW_LET@31..34 "let"
          WHITESPACE@34..35 " "
          IDENT@35..38 "arr"
          WHITESPACE@38..39 " "
          OP_ASSIGN@39..40 "="
          EXPR@40..75
            WHITESPACE@40..41 " "
            EXPR_ARRAY@41..75
              PUNCT_BRACKET_START@41..42 "["
              EXPR@42..43
                EXPR_LIT@42..43
                  LIT@42..43
                    LIT_INT@42..43 "1"
              PUNCT_COMMA@43..44 ","
              WHITESPACE@44..45 " "
              EXPR@45..49
                EXPR_LIT@45..49
                  LIT@45..49
                    LIT_BOOL@45..49 "true"
              PUNCT_COMMA@49..50 ","
              WHITESPACE@50..51 " "
              EXPR@51..58
                EXPR_LIT@51..58
                  LIT@51..58
                    LIT_FLOAT@51..58 "123.456"
              PUNCT_COMMA@58..59 ","
              WHITESPACE@59..60 " "
              EXPR@60..67
                EXPR_LIT@60..67
                  LIT@60..67
                    LIT_STR@60..67 "\"hello\""
              PUNCT_COMMA@67..68 ","
              WHITESPACE@68..69 " "
              EXPR@69..70
                EXPR_LIT@69..70
                  LIT@69..70
                    LIT_INT@69..70 "3"
              PUNCT_COMMA@70..71 ","
              WHITESPACE@71..72 " "
              EXPR@72..74
                EXPR_LIT@72..74
                  LIT@72..74
                    LIT_INT@72..74 "42"
              PUNCT_BRACKET_END@74..75 "]"
    PUNCT_SEMI@75..76 ";"
  WHITESPACE@76..78 "\n\n"
  STMT@78..417
    ITEM@78..259
      EXPR@78..259
        EXPR_FOR@78..259
          KW_FOR@78..81 "for"
          PAT@81..88
            WHITESPACE@81..82 " "
            PAT_TUPLE@82..88
              PUNCT_PAREN_START@82..83 "("
              PARAM@83..84
                IDENT@83..84 "a"
              PUNCT_COMMA@84..85 ","
              WHITESPACE@85..86 " "
              PARAM@86..87
                IDENT@86..87 "i"
              PUNCT_PAREN_END@87..88 ")"
          WHITESPACE@88..89 " "
          KW_IN@89..91 "in"
          EXPR@91..96
            WHITESPACE@91..92 " "
            EXPR_IDENT@92..96
              IDENT@92..95 "arr"
              WHITESPACE@95..96 " "
          EXPR_BLOCK@96..259
            PUNCT_BRACE_START@96..97 "{"
            WHITESPACE@97..102 "\n    "
            STMT@102..237
              ITEM@102..231
                EXPR@102..231
                  EXPR_FOR@102..231
                    KW_FOR@102..105 "for"
                    PAT@105..112
                      WHITESPACE@105..106 " "
                      PAT_TUPLE@106..112
                        PUNCT_PAREN_START@106..107 "("
                        PARAM@107..108
                          IDENT@107..108 "b"
                        PUNCT_COMMA@108..109 ","
                        WHITESPACE@109..110 " "
                        PARAM@110..111
                          IDENT@110..111 "j"
                        PUNCT_PAREN_END@111..112 ")"
                    WHITESPACE@112..113 " "
                    KW_IN@113..115 "in"
                    EXPR@115..144
                      WHITESPACE@115..116 " "
                      EXPR_ARRAY@116..143
                        PUNCT_BRACKET_START@116..117 "["
                        EXPR@117..120
                          EXPR_LIT@117..120
                            LIT@117..120
                              LIT_CHAR@117..120 "'x'"
                        PUNCT_COMMA@120..121 ","
                        WHITESPACE@121..122 " "
                        EXPR@122..124
                          EXPR_LIT@122..124
                            LIT@122..124
                              LIT_INT@122..124 "42"
                        PUNCT_COMMA@124..125 ","
                        WHITESPACE@125..126 " "
                        EXPR@126..128
                          EXPR_PAREN@126..128
                            PUNCT_PAREN_START@126..127 "("
                            PUNCT_PAREN_END@127..128 ")"
                        PUNCT_COMMA@128..129 ","
                        WHITESPACE@129..130 " "
                        EXPR@130..133
                          EXPR_LIT@130..133
                            LIT@130..133
                              LIT_INT@130..133 "123"
                        PUNCT_COMMA@133..134 ","
                        WHITESPACE@134..135 " "
                        EXPR@135..137
                          EXPR_LIT@135..137
                            LIT@135..137
                              LIT_INT@135..137 "99"
                        PUNCT_COMMA@137..138 ","
                        WHITESPACE@138..139 " "
                        EXPR@139..142
                          EXPR_LIT@139..142
                            LIT@139..142
                              LIT_FLOAT@139..142 "0.5"
                        PUNCT_BRACKET_END@142..143 "]"
                      WHITESPACE@143..144 " "
                    EXPR_BLOCK@144..231
                      PUNCT_BRACE_START@144..145 "{"
                      WHITESPACE@145..154 "\n        "
                      STMT@154..188
                        ITEM@154..188
                          EXPR@154..188
                            EXPR_IF@154..188
                              KW_IF@154..156 "if"
                              EXPR@156..165
                                EXPR_BINARY@156..165
                                  WHITESPACE@156..157 " "
                                  EXPR_IDENT@157..159
                                    IDENT@157..158 "b"
                                    WHITESPACE@158..159 " "
                                  OP_GT@159..160 ">"
                                  EXPR@160..165
                                    WHITESPACE@160..161 " "
                                    EXPR_LIT@161..164
                                      LIT@161..164
                                        LIT_INT@161..164 "100"
                                    WHITESPACE@164..165 " "
                              EXPR_BLOCK@165..178
                                PUNCT_BRACE_START@165..166 "{"
                                WHITESPACE@166..167 " "
                                STMT@167..176
                                  ITEM@167..175
                                    EXPR@167..175
                                      EXPR_CONTINUE@167..175
                                        KW_CONTINUE@167..175 "continue"
                                  PUNCT_SEMI@175..176 ";"
                                WHITESPACE@176..177 " "
                                PUNCT_BRACE_END@177..178 "}"
                              WHITESPACE@178..188 "\n\n        "
                      STMT@188..225
                        ITEM@188..224
                          EXPR@188..224
                            EXPR_CALL@188..224
                              EXPR@188..193
                                EXPR_IDENT@188..193
                                  IDENT@188..193 "print"
                              ARG_LIST@193..224
                                PUNCT_PAREN_START@193..194 "("
                                EXPR@194..223
                                  EXPR_LIT@194..223
                                    LIT@194..223
                                      LIT_STR@194..223 "`(${i}, ${j}) = (${a} ..."
                                PUNCT_PAREN_END@223..224 ")"
                        PUNCT_SEMI@224..225 ";"
                      WHITESPACE@225..230 "\n    "
                      PUNCT_BRACE_END@230..231 "}"
              WHITESPACE@231..237 "\n\n    "
            STMT@237..258
              ITEM@237..258
                EXPR@237..258
                  EXPR_IF@237..258
                    KW_IF@237..239 "if"
                    EXPR@239..247
                      EXPR_BINARY@239..247
                        WHITESPACE@239..240 " "
                        EXPR_IDENT@240..242
                          IDENT@240..241 "a"
                          WHITESPACE@241..242 " "
                        OP_EQ@242..244 "=="
                        EXPR@244..247
                          WHITESPACE@244..245 " "
                          EXPR_LIT@245..246
                            LIT@245..246
                              LIT_INT@245..246 "3"
                          WHITESPACE@246..247 " "
                    EXPR_BLOCK@247..257
                      PUNCT_BRACE_START@247..248 "{"
                      WHITESPACE@248..249 " "
                      STMT@249..255
                        ITEM@249..254
                          EXPR@249..254
                            EXPR_BREAK@249..254
                              KW_BREAK@249..254 "break"
                        PUNCT_SEMI@254..255 ";"
                      WHITESPACE@255..256 " "
                      PUNCT_BRACE_END@256..257 "}"
                    WHITESPACE@257..258 "\n"
            PUNCT_BRACE_END@258..259 "}"
    WHITESPACE@259..260 "\n"
    COMMENT_LINE@260..349 "//print(a);           ..."
    WHITESPACE@349..378 "\n                     ..."
    COMMENT_LINE@378..415 "//    because 'a' is  ..."
    WHITESPACE@415..417 "\n\n"
  STMT@417..486
    ITEM@417..485
      EXPR@417..485
        EXPR_FOR@417..485
          KW_FOR@417..420 "for"
          PAT@420..422
            WHITESPACE@420..421 " "
            PAT_IDENT@421..422
              IDENT@421..422 "i"
          WHITESPACE@422..423 " "
          KW_IN@423..425 "in"
          EXPR@425..442
            EXPR_CALL@425..441
              EXPR@425..431
                WHITESPACE@425..426 " "
                EXPR_IDENT@426..431
                  IDENT@426..431 "range"
              ARG_LIST@431..441
                PUNCT_PAREN_START@431..432 "("
                EXPR@432..433
                  EXPR_LIT@432..433
                    LIT@432..433
                      LIT_INT@432..433 "5"
                PUNCT_COMMA@433..434 ","
                WHITESPACE@434..435 " "
                EXPR@435..436
                  EXPR_LIT@435..436
                    LIT@435..436
                      LIT_INT@435..436 "0"
                PUNCT_COMMA@436..437 ","
                WHITESPACE@437..438 " "
                EXPR@438..440
                  EXPR_UNARY@438..440
                    OP_SUB@438..439 "-"
                    EXPR@439..440
                      EXPR_LIT@439..440
                        LIT@439..440
                          LIT_INT@439..440 "1"
                PUNCT_PAREN_END@440..441 ")"
            WHITESPACE@441..442 " "
          EXPR_BLOCK@442..485
            PUNCT_BRACE_START@442..443 "{"
            WHITESPACE@443..445 "  "
            COMMENT_LINE@445..469 "// runs from 5 down to 1"
            WHITESPACE@469..474 "\n    "
            STMT@474..483
              ITEM@474..482
                EXPR@474..482
                  EXPR_CALL@474..482
                    EXPR@474..479
                      EXPR_IDENT@474..479
                        IDENT@474..479 "print"
                    ARG_LIST@479..482
                      PUNCT_PAREN_START@479..480 "("
                      EXPR@480..481
                        EXPR_IDENT@480..481
                          IDENT@480..481 "i"
                      PUNCT_PAREN_END@481..482 ")"
              PUNCT_SEMI@482..483 ";"
            WHITESPACE@483..484 "\n"
            PUNCT_BRACE_END@484..485 "}"
    WHITESPACE@485..486 "\n"

